contracts/cw20-base/src/contract.rs

```diff
@@ -0,0 +1,280 @@
+use crate::error::ContractError;
+use crate::state::{MinterData, TokenInfo, BALANCES, LOGO, MARKETING_INFO, TOKEN_INFO};
+use crate::validation::{validate_accounts, validate_msg, verify_logo};
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{
+    Binary, Deps, DepsMut, Env, MessageInfo, Response, StdError, StdResult, Uint128,
+};
+use cw2::set_contract_version;
+use cw20::{
+    BalanceResponse, Cw20Coin, Cw20ReceiveMsg, Logo, LogoInfo, MarketingInfoResponse,
+    MinterResponse, TokenInfoResponse,
+};
+use sylvia::{contract, schemars};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cw20-base";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cw_serde]
+pub struct InstantiateMarketingInfo {
+    pub project: Option<String>,
+    pub description: Option<String>,
+    pub marketing: Option<String>,
+    pub logo: Option<Logo>,
+}
+
+#[cw_serde]
+pub struct InstantiateMsgData {
+    pub name: String,
+    pub symbol: String,
+    pub decimals: u8,
+    pub initial_balances: Vec<Cw20Coin>,
+    pub mint: Option<MinterResponse>,
+    pub marketing: Option<InstantiateMarketingInfo>,
+}
+
+pub struct Cw20Base {}
+
+#[contract]
+#[messages(cw20_allowances as Cw20Allowances)]
+impl Cw20Base {
+    pub const fn new() -> Self {
+        Self {}
+    }
+
+    pub fn create_accounts(
+        deps: &mut DepsMut,
+        accounts: &[Cw20Coin],
+    ) -> Result<Uint128, ContractError> {
+        validate_accounts(accounts)?;
+
+        let mut total_supply = Uint128::zero();
+        for row in accounts {
+            let address = deps.api.addr_validate(&row.address)?;
+            BALANCES.save(deps.storage, &address, &row.amount)?;
+            total_supply += row.amount;
+        }
+
+        Ok(total_supply)
+    }
+
+    #[msg(instantiate)]
+    pub fn instantiate(
+        &self,
+        ctx: (DepsMut, Env, MessageInfo),
+        data: InstantiateMsgData,
+    ) -> Result<Response, ContractError> {
+        let (mut deps, ..) = ctx;
+        let InstantiateMsgData {
+            name,
+            symbol,
+            decimals,
+            initial_balances,
+            mint,
+            marketing,
+        } = data;
+        set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+        // check valid token info
+        validate_msg(&name, &symbol, decimals)?;
+        // create initial accounts
+        let total_supply = Cw20Base::create_accounts(&mut deps, &initial_balances)?;
+
+        if let Some(limit) = mint.as_ref().and_then(|v| v.cap) {
+            if total_supply > limit {
+                return Err(StdError::generic_err("Initial supply greater than cap").into());
+            }
+        }
```

Comment:
You can collapse it to single if:

```rust
if matches!(
  Some(limit) if total_supply > limit,
  mint.as_ref().and_then(|v| v.cap)
) {
  return Err(...);
}
```

Also this is a perfect place for [`ensure`](https://docs.rs/cosmwasm-std/1.1.5/cosmwasm_std/macro.ensure.html):

```rust
ensure!(
  matches!(
    Some(limit) if total_supply > limit,
    mint.as_ref().and_then(|v| v.cap)
  ),
  Err(...)
);
```